patches/0117-libavfilter-dnn-dnn_backend_torch-Add-async-to-torch.patch

```diff
@@ -0,0 +1,127 @@
+index f54493d742..855b1b3b97 100644
+--- a/libavfilter/dnn/dnn_backend_torch.cpp
++++ b/libavfilter/dnn/dnn_backend_torch.cpp
+@@ -40,6 +40,8 @@ typedef struct THOptions{
+     char *device_name;
+     c10::DeviceType device_type;
+     int optimize;
++    uint8_t async;
++    uint32_t nireq;
+ } THOptions;
+ 
+ typedef struct THContext {
+@@ -76,6 +78,8 @@ typedef struct THRequestItem {
+ static const AVOption dnn_th_options[] = {
+     { "device", "device to run model", OFFSET(options.device_name), AV_OPT_TYPE_STRING, { .str = "cpu" }, 0, 0, FLAGS },
+     { "optimize", "turn on graph executor optimization", OFFSET(options.optimize), AV_OPT_TYPE_INT, { .i64 = 0 }, 0, 1, FLAGS},
++    { "nireq",           "number of request",             OFFSET(options.nireq),           AV_OPT_TYPE_INT,    { .i64 = 0 },     0, INT_MAX, FLAGS },
++    { "async",           "use DNN async inference",       OFFSET(options.async),           AV_OPT_TYPE_BOOL,   { .i64 = 1 },     0,       1, FLAGS },
+     { NULL }
+ };
+ 
+@@ -258,27 +262,40 @@ static DNNModel *dnn_load_model_th(const char *model_filename, DNNFunctionType f
+     }
+     first_param = *th_model->jit_model->named_parameters().begin();
+ 
++#if !HAVE_PTHREAD_CANCEL
++    if (ctx->options.async) {
++        ctx->options.async = 0;
++        av_log(filter_ctx, AV_LOG_WARNING, "pthread is not supported, roll back to sync.\n");
++    }
++#endif
++
+     th_model->request_queue = ff_safe_queue_create();
+     if (!th_model->request_queue) {
+         goto fail;
+     }
+ 
+-    item = (THRequestItem *)av_mallocz(sizeof(THRequestItem));
+-    if (!item) {
+-        goto fail;
+-    }
+-    item->lltask = NULL;
+-    item->infer_request = th_create_inference_request();
+-    if (!item->infer_request) {
+-        av_log(NULL, AV_LOG_ERROR, "Failed to allocate memory for Torch inference request\n");
+-        goto fail;
++    if (ctx->options.nireq <= 0) {
++        ctx->options.nireq = 1;
+     }
+-    item->exec_module.start_inference = &th_start_inference;
+-    item->exec_module.callback = &infer_completion_callback;
+-    item->exec_module.args = item;
+ 
+-    if (ff_safe_queue_push_back(th_model->request_queue, item) < 0) {
+-        goto fail;
++    for (int i = 0; i < ctx->options.nireq; i++) {
++        item = (THRequestItem *)av_mallocz(sizeof(THRequestItem));
++        if (!item) {
++            goto fail;
++        }
++        item->lltask = NULL;
++        item->infer_request = th_create_inference_request();
++        if (!item->infer_request) {
++            av_log(NULL, AV_LOG_ERROR, "Failed to allocate memory for Torch inference request\n");
++            goto fail;
++        }
++        item->exec_module.start_inference = &th_start_inference;
++        item->exec_module.callback = &infer_completion_callback;
++        item->exec_module.args = item;
++
++        if (ff_safe_queue_push_back(th_model->request_queue, item) < 0) {
++            goto fail;
```

Comment:
There might be memory leak in the error path because some items have been added into th_model->request_queue

Reply:
Good catch, Done. I also optimized the free logic.